api/bin/chainflip-ingress-egress-tracker/src/main.rs

```diff
@@ -1,17 +1,145 @@
-use chainflip_engine::settings::{
-	insert_command_line_option, CfSettings, HttpBasicAuthEndpoint, WsHttpEndpoints,
+use cf_chains::{
+	btc::BitcoinNetwork, evm::SchnorrVerificationComponents, AnyChain, Bitcoin, Chain, Ethereum,
+	Polkadot,
+};
+use cf_primitives::{Asset, BroadcastId, ForeignChain};
+use chainflip_engine::{
+	settings::{insert_command_line_option, CfSettings, HttpBasicAuthEndpoint, WsHttpEndpoints},
+	state_chain_observer::client::{
+		chain_api::ChainApi, storage_api::StorageApi, StateChainClient,
+	},
+	witness::common::STATE_CHAIN_CONNECTION,
 };
 use clap::Parser;
 use config::{Config, ConfigBuilder, ConfigError, Environment, Map, Source, Value};
 use futures::FutureExt;
 use jsonrpsee::{core::Error, server::ServerBuilder, RpcModule};
-use serde::Deserialize;
+use pallet_cf_broadcast::TransactionOutIdFor;
+use pallet_cf_ingress_egress::DepositWitness;
+use serde::{Deserialize, Serialize};
+use state_chain_runtime::PalletInstanceAlias;
 use std::{collections::HashMap, env, net::SocketAddr};
 use tracing::log;
-use utilities::task_scope;
+use utilities::{rpc::NumberOrHex, task_scope};
 
 mod witnessing;
 
+#[derive(Serialize)]
+struct WitnessAsset {
+	chain: ForeignChain,
+	asset: Asset,
+}
+
+impl From<cf_chains::assets::eth::Asset> for WitnessAsset {
+	fn from(asset: cf_chains::assets::eth::Asset) -> Self {
+		match asset {
+			cf_chains::assets::eth::Asset::Eth |
+			cf_chains::assets::eth::Asset::Flip |
+			cf_chains::assets::eth::Asset::Usdc =>
+				Self { chain: ForeignChain::Ethereum, asset: asset.into() },
+		}
+	}
+}
+
+impl From<cf_chains::assets::dot::Asset> for WitnessAsset {
+	fn from(asset: cf_chains::assets::dot::Asset) -> Self {
+		match asset {
+			cf_chains::assets::dot::Asset::Dot =>
+				Self { chain: ForeignChain::Polkadot, asset: asset.into() },
+		}
+	}
+}
+
+impl From<cf_chains::assets::btc::Asset> for WitnessAsset {
+	fn from(asset: cf_chains::assets::btc::Asset) -> Self {
+		match asset {
+			cf_chains::assets::btc::Asset::Btc =>
+				Self { chain: ForeignChain::Bitcoin, asset: asset.into() },
+		}
+	}
+}
+
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
```

Comment:
seems weird to call this `type` ? maybe just `id`?

Reply:
The enum tag i.e. `Bitcoin`, `Ethereum`, or `Polkadot` get added to the JSON representation under the field `type`, e.g.:

```json
{ "type": "bitcoin", "hash": "0x123456" }
```

Otherwise the default serialization is something like:
```json
{ "Bitcoin": { "hash": "0x123456" } }
```

Reply:
renamed the rename to `deposit_chain`

Reply:
yeah more the naming than anything but if you're happy, idm, not a strong opinion